ch_pipeline/analysis/spectral_lines.py

```diff
@@ -0,0 +1,253 @@
+        # to obtain the global index
+        outlier = [list(ind + off) for ind, off in zip(outlier_local, offset_local)]
+
+        # Combine the global index of outliers across all ranks
+        def concatenate_ranks(x, comm):
+            return np.array([r for rank in comm.allgather(x) for r in rank])
+
+        index = [concatenate_ranks(olist, self.comm) for olist in outlier]
+
+        # Extract the 3D coordinates corresponding to outliers
+        noutlier = index[0].size
+        raw_coord = np.zeros((noutlier, 3), dtype=float)
+        raw_coord[:, 0] = freq[index[0]]
+        raw_coord[:, 1] = ra[index[1]]
+        raw_coord[:, 2] = el[index[2]]
+
+        src_dec = np.arcsin(raw_coord[:, 2]) + np.radians(self.tel.latitude)
+
+        # Define the expected separation for a single source
+        max_ysep = self.max_ysep
+        freq_eval = raw_coord[:, 0]
+        if "beamform_ns_nsmax" in ringmap.attrs:
+            max_ysep = ringmap.attrs["beamform_ns_nsmax"]
+            if ringmap.attrs["beamform_ns_scaled"]:
+                freq_eval = ringmap.attrs["beamform_ns_freqmin"]
+
+        wavelength = scipy.constants.c / (freq_eval * 1e6)
+
+        delta = np.zeros((noutlier, 3), dtype=float)
+        delta[:, 0] = 1.0 / (2.35482 * self.delay_cutoff)
+        delta[:, 1] = cal_utils.guess_fwhm(freq_eval, pol="X", dec=src_dec, sigma=True)
+        delta[:, 2] = 0.85 * wavelength / (2.35482 * max_ysep)
+
+        coord = raw_coord / delta
+
+        # Compute pairwise distances and cluster
+        dists = pdist(coord)  # pairwise distances in rescaled space
+        linkage_matrix = linkage(dists, method="single")  # or 'complete', 'average'
+
+        cluster_ids = fcluster(linkage_matrix, t=self.nsigma, criterion="distance")
+
+        uniq_clusters = np.unique(cluster_ids)
+        nuniq = uniq_clusters.size
+
+        # Create output container
+        out = containers.SpectralLineCatalog(object_id=nuniq)
+
+        # Loop over clusters
+        for cc, cluster_id in enumerate(uniq_clusters):
+
+            # Identify which outliers are in this cluster,
+            # then identify which are held on this rank
+            this_group = np.flatnonzero(cluster_ids == cluster_id)
+            this_rank = np.array(
+                [gg for gg in this_group if index[0][gg] in local_freq_index]
+            )
+            local_npix = this_rank.size
+
+            # Check if we have cluster members on this rank
+            if local_npix > 0:
+
+                # Need to convert back to local indices to extract map values
+                local_ind = tuple(
+                    [ind[this_rank] - off for ind, off in zip(index, offset_local)]
+                )
+
+                map_values = rm[local_ind]
+                s2n_values = np.abs(map_values) * np.sqrt(wm[local_ind])
+
+                # Identify the maximum pixel within the local part of this cluster
+                imax = np.argmax(np.abs(map_values))
```

Comment:
I'm wondering if you'd want to instead select the pixel with the maximum SNR, since the pixel with the maximum absolute value could also have that value due to being very noisy. Not sure if I understood correctly. If what I'm thinking doesn't make sense, could you explain?

Reply:
Yeah that make sense.  I've updated the code to use the pixel with the maximum signal to noise.